src/System.IO.FileSystem/src/System/IO/Win32FileStreamAsyncResult.cs

```diff
@@ -96,232 +53,147 @@ unsafe private sealed class FileStreamAsyncResult : IAsyncResult
-                    // this will save us some perf, assuming we can correctly
-                    // initialize the ManualResetEvent.  
-                    if (_waitHandle == null)
-                    {
-                        ManualResetEvent mre = new ManualResetEvent(false);
-#if USE_OVERLAPPED
-                        if (_overlapped != null && _overlapped->EventHandle != IntPtr.Zero)
-                        {
-                            mre.SetSafeWaitHandle(new SafeWaitHandle(_overlapped->EventHandle, true));
-                        }
-#endif
-
-                        // make sure only one thread sets _waitHandle
-                        if (Interlocked.CompareExchange<ManualResetEvent>(ref _waitHandle, mre, null) == null)
-                        {
-                            if (_isComplete)
-                                _waitHandle.Set();
-                        }
-                        else
-                        {
-                            // There's a slight but acceptable race condition if 
-                            // we weren't the thread that set _waitHandle and 
-                            // this code path returns before the code in the if 
-                            // statement  executes (on the other thread). 
-                            // However, the caller is waiting for the wait 
-                            // handle to be set, which will still happen.
-                            mre.Dispose();
-                        }
-                    }
-                    return _waitHandle;
-                }
+                get { return ((IAsyncResult)_completionSource.Task).AsyncWaitHandle; }
             }
 
-            // Returns true iff the user callback was called by the thread that 
-            // called BeginRead or BeginWrite.  If we use an async delegate or
-            // threadpool thread internally, this will be false.  This is used
-            // by code to determine whether a successive call to BeginRead needs 
-            // to be done on their main thread or in their callback to avoid a
-            // stack overflow on many reads or writes.
-            public bool CompletedSynchronously
+            bool IAsyncResult.CompletedSynchronously
             {
-                get { return _completedSynchronously; }
+                get { return ((IAsyncResult)_completionSource.Task).CompletedSynchronously; }
             }
 
-            private void CallUserCallbackWorker()
+            public void SetCompletedSynchronously(int numBytes)
             {
-                _isComplete = true;
-
-                Interlocked.MemoryBarrier();
-
-                // ensure _isComplete is set before reading _waitHandle
-                if (_waitHandle != null)
-                    _waitHandle.Set();
-
-                _userCallback(this);
+                ReleaseNativeResource();
+                _completionSource.TrySetResult(numBytes + _numBufferedBytes);
```

Comment:
Ah - I missed that.
Added RunContinuationsAsynchronously to the TaskComplationSource's constructor and a comment to explain that it is for compat reasons.